### DISCUSSION TITLE
UX review of how the 'tree' structure is represented in the editor ui

### DISCUSSION THREAD

**Discussion by tmsndrs** [wagtail/wagtail#11382] [Ideas]
Hi, 

I know this is a fairly significant piece of work but it's come up from various users. I think it would be good to give some more consideration to how the content tree is represented in the editor ui. Editors often don't get enough sense of hierarchy and, especially on larger sites, it can be hard to have a sense of where you are on the tree.

Without interrupting existing workflows we could look at adding a different view for displaying content. 

As a very basic starting point I think it would be better if the arrow icon you click to navigate to child pages was an angled arrow like this one http://bit.ly/1N2keEo which would give a better sense of a tree structure rather than a flat model.

Not that I think this is the optimum solution for showing site structure but I've attached an of example of how Square space  do it. 

![screen shot 2015-04-07 at 11 03 00](https://cloud.githubusercontent.com/assets/7290588/7021137/ac0426b8-dd16-11e4-81b6-a9a110bdf4d0.png)

Cheers,

**Comment by JoshBarr** [comment]
Hey eveyone,

The RFC Google Group on this topic ate my reply, so I'm moving the conversation to somewhere more reliable!  

How's this for a hybrid of Tom comments and Brent's thinking so far? The sidebar could be used as a classic List/Detail view, showing the tree and all the nodes at the selected branch. This would let us easily scale up to 3000+ nodes via pagination, but also gives a sense of the hierarchy when you're editing. 

![ss-tree](https://cloud.githubusercontent.com/assets/1715456/7089647/806fd56e-dff1-11e4-94a2-2a3a3356f6de.jpg)

Also, as Brent mentioned, we've been tinkering with an alternative popover that's based on the OSX ColumnView interface.  

Demo video: [https://youtu.be/vLxoz4jy5UA](https://youtu.be/vLxoz4jy5UA) – watch it in 1080p for best effect.
Code:  [my fork](https://github.com/JoshBarr/wagtail/tree/feature/new-explorer/wagtail/wagtailadmin/static-src/js)

It's a proof of concept so nothing's serialised back to the server at this stage – I imagine that would be a fair bit of work :) 

cheers,

Josh

**Comment by davecranwell** [comment]
I like the way that popup replacement is done. Will consider it more.

That column interface is great too! Again some interesting ideas to consider. We're going to have to learn React and create some workflow for our javascript if we're to incorporate your contributions (assuming React is your framework/library of choice), so we'll discuss this in our next meeting.

**Comment by tmsndrs** [comment]
I think their are some aspects of this that look like they might work really well. I think it would be worth testing with some real site maps to see how well it scales. Some questions and comments:
- Would you see the columns scrolling individually or together? Seems like individually would work better. 
- Might it be helpful to have some way to collapse or further expand the columns to allow editors a more concentrated view? Just thinking, if I'm looking through a long list of blog posts I wouldn't want the additional visual cluster surrounding that list. 
- The idea of being able to edit the page title is nice in theory but consideration would need to be given to how this impacts on slug.  
- The modal for adding new pages is cool but how would it scale? We often have 20 or so content types. Additionally how would this deal with limiting subpage types?
- I've always liked the idea of some sort of notifications for when significant actions are successful (like moving a page). Some good inspiration here http://tympanus.net/Development/NotificationStyles/index.html
- Would it be possible to include some way of adding pages to the site nav from this ui without things being cluttered (though I still think some sort of menu builder would be better than the current solution).

**Comment by tmsndrs** [comment]
Actually, I think the best option to address my second point would be OS style buttons to show a list or columns. The columns view could even have additional page settings such as publish/unpublish date in the far right column. 
![screen shot 2015-04-15 at 22 03 51](https://cloud.githubusercontent.com/assets/7290588/7169260/b5676a8e-e3bb-11e4-8b9d-9faa3dd63d70.png)

**Comment by ollywillans** [comment]
Hi Josh,

I'm Olly I'm the Creative Director at Torchbox and I wanted to respond on this too. Sorry it's been a bit slow.

Firstly the _OSX style navigator_ is a great piece of work.

I think it's a really interesting design pattern for this problem, with some cool ideas thrown in. And really impressive to see it working too!

While we may not be able to implement it quickly because it raises quite a few backend issues like how we would create pages without validation, I definitely think we need to have it under consideration as a replacement pattern for the explorer.

The main reason I like it, is because it would be a great help with creating the initial structure and pages for a site. Which is something
that your user research has flagged as a need too.

As you know from the conversation in the Google Group, our current page explorer only shows 'section pages' which is what keeps it manageable for big sites.  And I'm a little unsure whether this approach of showing all section pages and pages, would be as good as just showing section pages.

If you have all pages and section pages in there, the pages could slow the editor down as they try to navigate down the tree - it will be more difficult to quickly select the relevant section page because you'd have to find them amongst the pages. We could offset this by having them all at the top or something although that would cause it's own problems.

Any thoughts?

In response to your hybrid of Tom and Brent's thinking:

I'm not totally sure about the sitemap representation in the left hand column. I'm not sure if this is necessary.

I think that Tom Ss' feedback was that users sometimes wondered where their pages are before they realise they have to click into a section on the Explorer widget. Generally, i'm quite keen that we make small tweaks where that's enough to fix an issue, and in this case, I think we could mitigate it by:
- Adding some extra descriptive labelling within each explorer pane about the sub-pages within it e.g showing [37 child pages] after the title (perhaps only on hover, on desktop so it doesn't clutter it too much).
- Creating more indicative icons for the arrow on the right that gives more of a sense of accessing a sub-section

![image](https://cloud.githubusercontent.com/assets/352835/7185362/212e8498-e45a-11e4-8f87-475634fb20fa.png)

We liked the way that you cleaned up the the main listing putting all that supporting info, moved it away from a more tabular look, without it really impacting on the sortable nature of the listing. Although this is currently quite similar to lots of other listings views, so it may be better to keep the general listing look (rather than the letterbox card type look) but reap the advantages of cleaning things up in the way that you have.

One of the issues that we considered when thinking how to make the structural nature of things more obvious in the page listing, is whether we could link the look at and feel of the Page listing more closely with the Explorer, for example making the icon to go down into the sub-section reminiscent of that same action on the explorer (as mentioned above), and perhaps bringing the same darker green colour hover that you get when you hover an item in the Explorer onto the whole row of the page listing.

Finally, in your mockup, I'm not quite clear which one the 'Section page' is? Is it the 'Home' page in this instance? Even though you have the section visible on the left, it still seems like this page should be always visible (although perhaps you'd argue that you can find it my selecting the right Page type filter?).

If the section page was always visible, how would you suggest representing it in the listing? Dave had it at the top of the column in his version, which is ok, but we need to consider what happens when you then filter. Is it sticky?

If you do have it at the top, you have the ability to showing the hierarchy by illustrating the other listing pages as children of it, which might be interesting (might be able to combine a page icon for each of these with a colour showing publish status).

Hope to hear what you think of all of this and whether you'd be happy to do another iteration?

Cheers,

Olly

**Comment by davecranwell** [comment]
![image](https://cloud.githubusercontent.com/assets/352835/7185440/7f5a6046-e45a-11e4-9f93-d950204c9478.png)

That image which doesn't seem to be working in Olly's post

**Comment by JoshBarr** [comment]
Hey guys,

There's a lot to think about here! I took a stab at quickly implementing some different nav structures atop the CMS page tree... hacking the `explorer_menu` endpoint to return JSON instead of HTML so we can play with real world data. 

[Video](https://youtu.be/ALXTpmQLE_U)

With all these kinds of explorer, there's some confusion about what happens when you click on a node. In a desktop file browser you'd expect it to show the children and information about the node, and any actions you wanted to take would be a separate step afterwards. In conventional web wisdom, clicking a hyperlink _should_ take me directly to the resource. Tricky one. 

I agree it's unclear in my mockup above which node you're actually inside. It's the homepage, and we're seeing its children (i think). 

Windows explorer's tree view doesn't include files in the list (only directories). But it does feel a little strange not seeing any child nodes. Maybe you could show the first 10 children for each node, and hide the rest behind a "see all..." link? Doesn't feel that great though..

The idea of merging the page view and the explorer so they feel more connected is great. That will result in less interface on the whole and a more natural 'content-as-navigation' journey.  Brent and I will keep plugging away at some things and prototyping them based on your thoughts above. It must be possible to simplify this a bunch (said everyone who ever tried to build a CMS!). 

Since a tree will always be a tree, another interesting idea might be to soup up the other entry points to the content. Explore navigating based on tags/categories, or promoting searching (with really good autocomplete) as a way of discovering things – a kind of Filtered Search (a bit like Google Drive or even Github issues). "Show me all StandardPages filtered by the tag 'How-to' and the status 'draft', written by Jane". "Show me all content scheduled to go live next Tuesday". "Show me all revisions of a given page". 

By giving users a bunch of other tools to find the node they're looking for, it might take some of the pressure off the explorer tree?

**Comment by SalahAdDin** [comment]
So wonderful, :+1:

**Comment by ollywillans** [comment]
Hi Josh,

Great stuff! It's really nice to be able to switch Explorer views.​ 

Perhaps it could be a configuration thing to determine whether people see child pages or not.

I think the blue highlighting works ok to show the path and the active section. The key thing to me seems to be about having a clear way of ​showing users how to see the pages within a section when they want to. Maybe this is about linking off to the listing page as currently happens, or maybe there is some way of showing it within the explorer, see attached quick sketch. Even then, how do you get to edit that section page (not shown in my sketch)...

With respect to giving users another non-Explorer way to find a page:

That was definitely in our original thinking. We envisaged having the tree interface and a search interface with lots of filters (e.g. find draft pages by author x, of type y ordered by date created). We never built it though and haven't really missed it badly yet, perhaps because the search is pretty good.

This interface would still be useful though, perhaps not so much for editors looking for pages to edit (search does that, as does the frontend 'Edit in Wagtail' bird), but for admins wanting to 'check through all draft pages' or 'find all pages by author x' etc.

Thanks to you and Brent for keeping going on this. We're loving it.

Olly

![editwagtail](https://cloud.githubusercontent.com/assets/1789475/7346754/6dc98a96-ecd9-11e4-9db3-9238d1f71c0a.png)

**Comment by JoshBarr** [comment]
Hey that's a neat idea, being able to toggle between browsing the hierarchy and working with the children.

This is pretty hack-ish but demonstrates how that might feel. [Video](http://youtu.be/fRo4YHdsQp4)

**Comment by SalahAdDin** [comment]
It's very, very good!!!
I like it.
I think you could put a different skins for the admin? at least different colors :D.
So, in the last image i see a dark blue, but, i think that clash with the colors of the template.

**Comment by MechanisM** [comment]
Look at this one http://tympanus.net/Development/MultiLevelPushMenu/

**Comment by simonmaris** [comment]
I do agree that the current UX of the explorer could be improved (running 1.0 here..).

For example I find myself very often just clicking on the title of an element (that is just a container, e.g for blog posts). This takes me to the editing screen (which is consistent obviously). However, as I know there is nothing in there I would need to edit I was just expecting it to take me to its children.

So, my point is that the arrow all the way on the right (which gets highlighted once I hover an item) is just to far away and only noticeable (highlighted) after I already moved my cursor to the title. I know this is something I just have to get used to but I'm not sure it should be like that.. Mostly when I use the explorer I need it to move down through the tree quite quickly to find a document I want to work on. This is where I will spend my time. 

Looking at Josh's video (columnview) I feel like the separation between editing actions (opening a document) and navigation actions is much clearer and therefore improves the workflow. I can move down the tree quickly and then one additional click will let me edit the document. I think this is already an improvement on the current system and as far as I understand something like this should be fairly easy to implement without backend problems (as long as there is no dynamic add/delete etc - please correct me if I'm wrong)?

With respect to sites with a very large tree I do indeed think that the display of children in the tree should be a configuration option. After all, for most pages under let's say 50 pages, the display of children just makes finding the page you actually want to go to so much quicker. 

I do also like the MultiLevelPushMenu MechanisM linked to. I think it deals with screen estate in a very elegant way and feels easy and intuitive to handle. Maybe some overlapping could indeed be a good idea (unlike osx finders columnview).

Looking forward to hear from others, I wonder whether anything happened on this since April..

**Comment by alexgleason** [comment]
Nice!! Any updates on this?

**Comment by SalahAdDin** [comment]
I ask the same!

**Comment by davecranwell** [comment]
We're in the planning stages of some general UX testing of Wagtail (lead by @tmsndrs, the OP), including the explorer, from which we hope to get some direction on what is a priority in terms of UX improvements.

Parallel to this the developers have also been discussing a general update to the technology behind key areas of the UI notably the modals and explorer popout perhaps/probably using React. This is probably a precursor to being able to update the UX of the explorer or the tree paradigm and is still currently just an idea.

So things are in motion, but I wouldn't want to speculate on when there will be something to see.

**Comment by ramonakira** [comment]
Is there news to share on this?

**Comment by spaceghost-nl** [comment]
**Hi all. Me and Thijs have worked on the future of the explorer or now called pages. Here are teh results of 2 days of sprinting.** 

Inspired by this youtube video https://www.youtube.com/watch?v=vLxoz4jy5UA and talking with Olly an Ben from Torchbox, Thijs and me started thinking about what we could improve to this design so you only need the explorer as pagenavigator and we can remove all others ways of page navigation.

Things we loved about the design in the video;
On click see your page option (edit, delete, add) and also at the same time show all childpages (if there are any).
Collapsing of the first column if there are more than 2 columns visible (as you can see in example 2)

The changes we made in comparising to the video:

Change page options “Preview, Edit, Delete, Add” to “Edit, Copy, View live”.

The reason behind the changes are that:
Preview is mostly only done after you edited a page, so its more logical to put it on the editable page.
Delete is a hard action, You usually don’t wanna do that if you can’t see the content of the page. So also better placed at the editable page.
In the video “add” adds a page in the next column. This feels kinda weird. We thought it’s better to place “Add page” button in the column where you want to place the new page.
Copy is a debatable option to have but if you want to have it, here is a good place to put it. This way you could easily copy an existing page.
Instead of preview, here it is more expected that a user wants to see how a page looks live. So with the View live button this is easily achieved.

In the video we also see that every page has a round indicator before the name. This shows if the page is published or not. Assuming that the user knows every page is published we felt this indicator was not necessary. In stead we only show when a page is not published or is a draft by showing the label in the designs. This makes the design seem more at ease.

Another issue we had to solve was that, if a column has a lot of pages in it, there has to be a way to search or filter these pages. So our idea was that when a column has more than, say 25 pages, the column will have a menu at the top with (see design 2):

1. A Sorting option ( menu will collapse on click)
2. A search option (searchfield will expand on click)
3. Expand option. This wil expand the column to a full page view with more filtering options. Most commonly this will be used with columns that have a huge amount of pages.


![wt menu 2 column](https://cloud.githubusercontent.com/assets/16794514/24355381/ff648408-12f5-11e7-8c0a-0b3d69aed3e4.jpg)
![wt menu 3 column](https://cloud.githubusercontent.com/assets/16794514/24355382/ff66b836-12f5-11e7-81b3-a8621e13de16.jpg)
